### PR TITLE
Use new `ListResources` in the webapi layer for pagination and filtering

### PR DESCRIPTION
**Pull request by kimlisa**
part of [RFD 55](https://github.com/gravitational/teleport/blob/master/rfd%2F0055-webui-ss-paginate-filter.md#client-web-ui)

#### Description
In addition proccess query params sent in with request

#### Before Merge
-  [ ] Requires webassets update with corresponding changes (or else the web UI list views will be inaccurate)
- [ ] Mark RFD 55 state from draft to implemented

### CODE REVIEW

File: lib/web/resources.go [hunk 1]
```diff
@@ -288,6 +291,44 @@ func ExtractResourceAndValidate(yaml string) (*services.UnknownResource, error)
...
+
+	limit, err := queryLimit(values, "limit", defaults.MaxIterationLimit)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Sort is expected in format `<fieldName>:<asc|desc>` where
+	// index 0 is fieldName and index 1 is direction.
+	// If a direction is not set, or is not recognized, it defaults to ASC.
+	sortBy := types.SortBy{}
```
Comment (rosstimothy) [inline lib/web/resources.go:306]:
```suggestion
	var sortBy types.SortBy
```

File: lib/web/ui/server.go [hunk 2]
```diff
@@ -101,38 +101,28 @@ func MakeServers(clusterName string, servers []types.Server) []Server {
...
-			kubeClusters[cluster.Name] = cluster
-		}
-	}
-
-	uiKubeClusters := make([]Kube, 0, len(kubeClusters))
-	for _, cluster := range kubeClusters {
+func MakeKubeClusters(clusters []types.KubeCluster) []KubeCluster {
+	uiKubeClusters := make([]KubeCluster, 0, len(clusters))
+	for _, cluster := range clusters {
 		uiLabels := []Label{}
```
Comment (rosstimothy) [inline lib/web/ui/server.go:116]: You could allocate the capacity for `uiLabels`
```suggestion
		staticLabels := cluster.GetStaticLabels()
		dynamicLabels := cluster.GetDynamicLabels()
		uiLabels := make([]Label, 0, len(staticLabels) + len(dynamicLabels)
```

File: lib/web/apiserver_test.go [hunk 3]
```diff
@@ -536,7 +536,7 @@ func (s *WebSuite) TestSAMLSuccess(c *C) {
 
 	err = s.server.Auth().CreateSAMLConnector(connector)
 	c.Assert(err, IsNil)
-	s.server.Auth().SetClock(clockwork.NewFakeClockAt(time.Date(2017, 05, 10, 18, 53, 0, 0, time.UTC)))
+	s.server.Auth().SetClock(clockwork.NewFakeClockAt(time.Date(2017, 0o5, 10, 18, 53, 0, 0, time.UTC)))
```
Comment (rosstimothy) [inline lib/web/apiserver_test.go:539]: Was this change accidental?

File: lib/web/resources_test.go [hunk 4]
```diff
@@ -331,6 +334,104 @@ version: v2`
...
+		expected        proto.ListResourcesRequest
+	}{
+		{
+			name: "decode complex query correctly",
+			url:  "https://dev:3080/login?query=(labels%5B%60%22test%22%60%5D%20%3D%3D%20%22%2B%3A'%2C%23*~%25%5E%22%20%26%26%20!exists(labels.tier))%20%7C%7C%20resource.spec.description%20!%3D%20%22weird%20example%20https%3A%2F%2Ffoo.dev%3A3080%3Fbar%3Da%2Cb%26baz%3Dbanana%22",
+			expected: proto.ListResourcesRequest{
+				ResourceType:        types.KindNode,
+				Limit:               defaults.MaxIterationLimit,
+				NeedTotalCount:      true,
+				SearchKeywords:      []string{},
```
Comment (rosstimothy) [inline lib/web/resources_test.go:353]: You should be able to use the default value here and below
```suggestion
```

File: lib/web/resources_test.go [hunk 5]
```diff
@@ -331,6 +334,104 @@ version: v2`
...
+			m := &mockedResourceAPIGetter{}
+			m.mockListResources = func(ctx context.Context, req proto.ListResourcesRequest) (*types.ListResourcesResponse, error) {
+				if !tc.wantBadParamErr {
+					require.Equal(t, tc.expected, req)
+				}
+				return nil, nil
+			}
+
+			_, err = listResources(m, httpReq, types.KindNode)
+			switch tc.wantBadParamErr {
```
Comment (rosstimothy) [inline lib/web/resources_test.go:425]: Why not just an if/else?

**Review by rosstimothy — APPROVED** [review]

File: lib/web/apiserver_test.go [hunk 6]
```diff
@@ -772,6 +772,7 @@ func TestClusterNodesGet(t *testing.T) {
 	nodes := clusterNodesGetResponse{}
 	require.NoError(t, json.Unmarshal(re.Bytes(), &nodes))
 	require.Len(t, nodes.Items, 1)
+	require.Equal(t, nodes.TotalCount, 1)
```
Comment (zmb3) [inline lib/web/apiserver_test.go:775]: Technically, the signature is `require.Equal(t, EXPECTED, ACTUAL)`.

We've switched the expected/actual values, which is fine if the test passes, but can result in a confusing error message if there's a value.

File: lib/web/apiserver_test.go [hunk 7]
```diff
@@ -2001,6 +2003,7 @@ func TestClusterDatabasesGet(t *testing.T) {
 	resp = testResponse{}
 	require.NoError(t, json.Unmarshal(re.Bytes(), &resp))
 	require.Len(t, resp.Items, 1)
+	require.Equal(t, resp.TotalCount, 1)
```
Comment (zmb3) [inline lib/web/apiserver_test.go:2006]: Same comment as above. And this occurs a few more times below. I would just take a pass and check all new assertions.

**Review by zmb3 — APPROVED** [review]
Just a couple minor comments.

**Review by espadolini — APPROVED** [review]
Nothing to add

**Review by russjones — APPROVED** [review]

**Comment by kimlisa** [comment]
May i get a review for this commit please from either @zmb3, @espadolini, or @rosstimothy: https://github.com/gravitational/teleport/pull/11019/commits/584feaa909abb59e003c11de25e1e017b22c82ff

The web UI needed a flag that would tell it if the list is empty from filter results, or if user just didn't have any resource to access. These are the views depending on this flag and list results:

1) If a list result was empty, and user had no resources to access, and the user has `read + write` access then the UI will show the user `how to get started`:
image

2) If a list result was empty, and user had no resources to acces, and the user has `read only` access then we show this screen:
image

3) If a list result was empty, and user HAS resources to access, then it means there was no results found after applying filter:
image

**Comment by espadolini** [comment]
@kimlisa in a hypothetical future in which we can do filtering in the backend itself that extra flag would force us to still figure out if something else could've been available - wouldn't it be simpler to only show the getting started UI element if the filter textbox is empty?

**Comment by kimlisa** [comment]
> @kimlisa in a hypothetical future in which we can do filtering in the backend itself that extra flag would force us to still figure out if something else could've been available - wouldn't it be simpler to only show the getting started UI element if the filter textbox is empty?

i think you're right @espadolini, it's also weird that we are ignoring the filter field too (in the UI b/c it will get replaced with the empty state)

**Pull request merged**